proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,135 @@
+
+  // The following fields contain in-toto artifact rules identifying the
+  // artifacts that enter this supply chain step, and exit the supply chain
+  // step, i.e. materials and products of the step.
+  repeated google.protobuf.ListValue expected_materials = 3;
+  repeated google.protobuf.ListValue expected_products = 4;
+
+  // This field contains the expected command used to perform the step.
+  repeated string expected_command = 5;
+
+  // This field contains a value that indicates the minimum number of keys that
+  // need to be used to sign the step's in-toto attestation.
+  int64 threshold = 6;
+}
+
+// This defines the format used to record keys used in the software supply
+// chain. An in-toto link is attested using one or more keys defined in the
+// in-toto layout.
+message SigningKey {
+  // key_id is an identifier for the signing key.
+  string key_id = 1;
+
+  // This field identifies the hash algorithms used in the particular key ID.
+  repeated string key_id_hash_algorithms = 2;
+
+  // This field identifies the specific signing method. Eg: "RSASSA-PSS",
+  // "ed25519", and "ecdsa".
+  string key_type = 3;
+
+  // This field contains the actual public key.
+  string public_key_value = 4;
+
+  // This field contains the overall scheme used for the creation of the signing
+  // key. Eg: "rsassa-pss-sha256".
+  string key_scheme = 5;
+}
+
+// This corresponds to a signed in-toto attestation - it is made up of one or
+// more signatures and the link attestation itself. This is used for occurrences
+// of a Grafeas in-toto note.
+message Details {
+  repeated Signature signatures = 1;
+  Link link = 2 [json_name="signed"];
+}
+
+// A signature object consists of the KeyID used and the signature itself.
+message Signature {
+  string key_id = 1 [json_name="keyid"];
+  string signature = 2 [json_name="sig"];
+}
+
+// This corresponds to an in-toto link attestation.
+message Link {
+  // This field contains the full command executed for the step. This can also
+  // be empty if attestations are generated for operations that aren't directly
+  // mapped to a specific command. Each term in the command is an independent
+  // string in the list. An example of a command in the in-toto metadata field
+  // is:
+  // "command": ["git", "clone", "https://github.com/in-toto/demo-project.git"]
+  repeated string effective_command = 1 [json_name="command"];
+
+  // Materials are the supply chain artifacts that go into the step and are used
+  // for the operation performed. The key of the map is the path of the artifact
+  // and the structure contains the recorded hash information. An example is:
+  // "materials": {
+  //   "foo/bar": {
+  //     "sha256": "ebebf..."
+  //   }
+  // }
+  map<string, google.protobuf.Struct> materials = 2;
+
+  // Products are the supply chain artifacts generated as a result of the step.
+  // The structure is identical to that of materials.
+  map<string, google.protobuf.Struct> products = 3;
```

Comment:
ditto here for the `Material`; you could also define another type, `Product`, in case you ever expect the definitions to diverge.

Reply:
merging into 1 thread, resolving here